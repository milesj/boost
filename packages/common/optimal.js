module.exports = require('./lib/optimal.js');
