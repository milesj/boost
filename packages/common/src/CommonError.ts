import { createScopedError } from '@boost/internal';

const errors = {
	PARSE_INVALID_EXT: 'Unable to parse file "{0}". Unsupported file extension.',
	PATH_REQUIRE_ABSOLUTE: 'An absolute file path is required.',
	PATH_RESOLVE_LOOKUPS: 'Failed to resolve a path using the following lookups (in order):\n{0}\n',
	PROJECT_NO_PACKAGE: 'No `package.json` found within project root.',
};

export type CommonErrorCode = keyof typeof errors;

export default createScopedError<CommonErrorCode>('CMN', 'CommonError', errors);
