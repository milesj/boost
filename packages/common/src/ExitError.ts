export class ExitError extends Error {
	code: number;

	constructor(message: string, code: number) {
		super(message);

		this.code = code;
		this.name = 'ExitError';
	}
}
