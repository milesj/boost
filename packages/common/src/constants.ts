// https://github.com/npm/validate-npm-package-name
export const MODULE_NAME_PART = /[a-z0-9]{1}[-a-z0-9_.]*/u;

export const MODULE_NAME_PATTERN = new RegExp(
	`^(?:@(${MODULE_NAME_PART.source})/)?(${MODULE_NAME_PART.source})$`,
	'u',
);
