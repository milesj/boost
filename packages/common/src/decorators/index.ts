export { default as Bind } from './Bind';
export { default as Debounce } from './Debounce';
export { default as Deprecate } from './Deprecate';
