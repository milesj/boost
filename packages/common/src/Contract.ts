import optimal, { Blueprint, Predicates, predicates } from 'optimal';
import { Optionable } from './types';

export default abstract class Contract<T extends object = {}> implements Optionable<T> {
  readonly options: Readonly<Required<T>>;

  constructor(options?: T) {
    this.options = this.configure(options);
  }

  /**
   * Set an options object by merging the new partial and existing options
   * with the defined blueprint, while running all validation checks.
   * Freeze and return the options object.
   */
  configure(options?: Partial<T> | ((options: Required<T>) => Partial<T>)): Readonly<Required<T>> {
    const nextOptions = typeof options === 'function' ? options(this.options) : options;

    // We don't want the options property to be modified directly,
    // so it's read only, but we still want to modify it with this function.
    // @ts-expect-error
    this.options = Object.freeze(
      optimal(
        { ...this.options, ...nextOptions },
        this.blueprint(predicates, this.options === undefined) as Blueprint<T>,
        {
          name: this.constructor.name,
        },
      ),
    );

    return this.options;
  }

  /**
   * Define an optimal blueprint in which to validate and build the
   * options object passed to the constructor, or when manual setting.
   */
  abstract blueprint(predicates: Predicates, onConstruction?: boolean): Blueprint<object>;
}
