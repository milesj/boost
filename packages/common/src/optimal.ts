import { instance, shape, string, union } from 'optimal';
import { Path } from './Path';
import { PortablePath } from './types';

export * from 'optimal';

/**
 * A schema for validating a value is a `PortablePath`.
 * Checks for a string, `Path`, or `Pathable`.
 */
export const portablePathSchema = union<PortablePath>('').of([
	string(),
	instance().of(Path),
	instance().of(Path, true),
	shape({
		path: string(),
	}),
]);
