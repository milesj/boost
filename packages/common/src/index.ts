/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal, { predicates, Blueprint, Predicates } from 'optimal';
import CommonError, { CommonErrorCode } from './CommonError';
import Contract from './Contract';
import ExitError from './ExitError';
import Path from './Path';
import PathResolver from './PathResolver';
import Project from './Project';
import * as json from './serializers/json';
import * as yaml from './serializers/yaml';

export * from './constants';
export * from './helpers';
export * from './types';

export {
  CommonError,
  CommonErrorCode,
  Contract,
  Blueprint,
  ExitError,
  Path,
  PathResolver,
  Predicates,
  Project,
  optimal,
  predicates,
  json,
  yaml,
};
