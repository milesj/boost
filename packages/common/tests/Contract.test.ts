import { Contract } from '../src';
import { Schemas } from '../src/optimal';

describe('Contract', () => {
	class OptionalProps extends Contract<{ foo?: string; bar?: number; baz?: { qux: string } }> {
		blueprint({ number, string, shape }: Schemas) {
			return {
				foo: string('default'),
				bar: number(),
				baz: shape({ qux: string() }),
			};
		}
	}

	class RequiredProps extends Contract<{ bar: number }> {
		constructor(options: { bar: number }) {
			super(options);
		}

		blueprint({ number }: Schemas) {
			return {
				bar: number().required(),
			};
		}
	}

	describe('constructor()', () => {
		it('errors for invalid option type passed', () => {
			expect(
				() =>
					new OptionalProps({
						// @ts-expect-error Invalid type
						foo: 123,
					}),
			).toThrowErrorMatchingSnapshot();
		});

		it('errors for unknown option', () => {
			expect(
				() =>
					new OptionalProps({
						// @ts-expect-error Unknown arg
						unknown: 'abc',
					}),
			).toThrowErrorMatchingSnapshot();
		});

		it('sets options', () => {
			const opts = new OptionalProps({
				foo: 'custom',
			});

			expect(opts.options.foo).toBe('custom');
		});

		it('inherits option defaults', () => {
			const opts = new OptionalProps();

			expect(opts.options.foo).toBe('default');
		});

		it('requires non-optional options to be passed', () => {
			expect(
				() =>
					// @ts-expect-error Missing arg
					new RequiredProps(),
			).toThrowErrorMatchingSnapshot();
		});
	});

	describe('configure()', () => {
		let opts: OptionalProps;

		beforeEach(() => {
			opts = new OptionalProps();
		});

		it('errors for invalid option type passed', () => {
			expect(() => {
				opts.configure({
					// @ts-expect-error Invalid type
					foo: 123,
				});
			}).toThrowErrorMatchingSnapshot();
		});

		it('errors for unknown option', () => {
			expect(() => {
				opts.configure({
					// @ts-expect-error Unknown arg
					unknown: 'abc',
				});
			}).toThrowErrorMatchingSnapshot();
		});

		it('can set an option', () => {
			expect(opts.options.foo).toBe('default');

			opts.configure({
				foo: 'new',
			});

			expect(opts.options.foo).toBe('new');
		});

		it('can set an option using a callback function', () => {
			expect(opts.options.foo).toBe('default');

			opts.configure((prev) => ({
				foo: `${prev.foo}-new`,
			}));

			expect(opts.options.foo).toBe('default-new');
		});

		it('persists other options not being configured', () => {
			opts = new OptionalProps({
				foo: 'abc',
				bar: 123,
			});

			expect(opts.options).toEqual({
				foo: 'abc',
				bar: 123,
				baz: { qux: '' },
			});

			opts.configure({
				bar: 456,
			});

			expect(opts.options).toEqual({
				foo: 'abc',
				bar: 456,
				baz: { qux: '' },
			});
		});

		it('freezes the object', () => {
			expect(() => {
				// @ts-expect-error Invalid type
				opts.options.foo = 'override';
			}).toThrowErrorMatchingSnapshot();
		});
	});

	describe('blueprint()', () => {
		const spy = jest.fn();

		class BlueprintTest extends Contract<{}> {
			blueprint(schemas: unknown, onConstruct: boolean) {
				spy(onConstruct);

				return {};
			}
		}

		it('passes true for 2nd arg only on construction', () => {
			const test = new BlueprintTest();

			expect(spy).toHaveBeenCalledWith(true);

			spy.mockClear();

			test.configure({});

			expect(spy).toHaveBeenCalledWith(false);
		});
	});
});
