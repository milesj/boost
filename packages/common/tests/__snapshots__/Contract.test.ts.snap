// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract configure() errors for invalid option type passed 1`] = `
"The following validations have failed for \`OptionalProps\`:
  - Invalid field \\"foo\\". Must be a string, received number."
`;

exports[`Contract configure() errors for unknown option 1`] = `
"The following validations have failed for \`OptionalProps\`:
  - Unknown fields: unknown."
`;

exports[`Contract configure() freezes the object 1`] = `"Cannot assign to read only property 'foo' of object '#<Object>'"`;

exports[`Contract constructor() errors for invalid option type passed 1`] = `
"The following validations have failed for \`OptionalProps\`:
  - Invalid field \\"foo\\". Must be a string, received number."
`;

exports[`Contract constructor() errors for unknown option 1`] = `
"The following validations have failed for \`OptionalProps\`:
  - Unknown fields: unknown."
`;

exports[`Contract constructor() requires non-optional options to be passed 1`] = `
"The following validations have failed for \`RequiredProps\`:
  - Invalid field \\"bar\\". Field is required and must be defined."
`;
